README.md
GSvarWeb
----------
[![Build Status](https://travis-ci.org/imgag/GSvarWeb.svg?branch=master)](https://travis-ci.org/imgag/GSvarWeb)

`GSvarWeb` is a remote API and browser application for [ngs-bits](https://github.com/imgag/ngs-bits).

Currently it is capable of displaying and processing `.GSvar` files

![GSvar table](./screenshots/gsvar-table.png "GSvar table")

## Architecture

`GSvarWeb` consists of server API which is [specified](./openapi.yaml) and implemented with [flask](http://flask.pocoo.org) using [OpenAPI](https://www.openapis.org/). 

Installation instructions can be found in the [API folder](./api/README.md).

The following environment variables can be set during startup:

| Variable        | Meaning                                    | Default |
| --------------- |:------------------------------------------:|:-------:|
| PORT            | Which port the server should listen on     | 8080    |
| DATA            | Where to put uploaded files                | $PWD    |
| NGS_BITS        | Where to find ngs bits binaries            | $PWD    |
| MEGSAP          | Where to find the megSAP directory         | None    |
| PRODUCTION      | Whether or not the server is in production | False   |
| AUTH_DOMAIN     | The domain to authenticate against  | auth.imgag.de  |
| AUTH_REALM      | The authentication realm                   | debug   |
| CORS_ORIGINS    | Comma-seperated list of allowed hosts      | -       |

You can start the server using above variables like so:

```
DATA=$PWD/data NGS_BITS=/some/path/to/ngs-bits/bin python3 -m openapi_server
```

You can find the OpenAPI UI under `/v1/ui` for testing and experiment purposes.

For convenience reasons the server also can serve files from the `dist` folder, allowing it to serve the frontend.

The API works by invoking shell pipelines and the command line interface of `ngs-bits`.

### Frontend

The frontend is a single page application built with [Vue](https://vuejs.org/) and [Vuetify](https://vuetifyjs.com/en/). Build instructions can be found in [frontend folder](./frontend/README.md).

The following environment variables can be set:

| Variable          | Meaning                                                  | Default                  | 
| ----------------- | -------------------------------------------------------- | ------------------------ |
| VUE_APP_API_URL   | Which URL to use for API requests                        | http://localhost:9000/v1 |
| VUE_APP_REALM     | The authorization realm to use                           | debug                    |

You can invoke environment variables like so:

```
VUE_APP_API_URL=http://localhost:9000/v1  npm run build
```

After building the project one can copy the `dist` folder into the `api` directory. When setting `SERVE_DIST`, it will be displayed as the default.

## Deployment in production

Currently `GSvarWeb` has not been tested in a production environment (meaning, it has not been publicly deployed to the WWW).

It is strongly recommended to use standalone [WSGI](https://wsgi.readthedocs.io/en/latest/index.html) containers, [such as gunicorn](http://flask.pocoo.org/docs/1.0/deploying/wsgi-standalone/#gunicorn), additionally behind a proxy.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 0.0.6 - 29.07.2019
### Added
- Make it possible to rate variants yourself in the frontend
- Display gchboc rating in table

### Changed
- Dont use ratings - use classifications @marc-sturm
- Changed title to GSVarWeb for GCHBOC
- Correctly insert records so they will not be overriden if they arent the same
- Automatically reload annotated file when rating
- Always download annotated file

## 0.0.5 - 29.07.2019
### Added
- Added controllers for rating variants.
- Added color for ClinVar and HGMD in ColumnDialog
- Add tooltip to ClinVar and HGMD
- Added Google link
- Added GnomAD link

### Changed
- Fix coloring at least for HGMD and ClinVar
- Make snackbars display permanently
- Remove caching everywhere

## 0.0.4
### Added
- Added logout button
- Add tooltip for all cells

### Changed
- Add development server to default CORS_ORIGIN
- Made ColumnDialog subheadings -> titles
- Split COSMIC entries by space
- Use subprocess to catch errors when filtering
- Split CORS_ORIGIN env variable

## 0.0.3
### Added
- Added download button
- Added netlify support

### Changed
- Rename REALM -> AUTH_REALM
- Rename NGS_BITS_BIN -> NGS_BITS
- Rename MEGSAP_DIR -> MEGSAP
- Rename VUE_APP_BASEPATH -> VUE_APP_API_URL
- Rename NGS_BITS_DATA -> DATA
- Change way of handling origins variable
- Removed v-data-table with ag-grid
- Rename GSVar -> GSvar everywhere
- Split GeneSplicer by &
- Add better error message when filter fails
- Break COSMIC into multiple links using ,
- Changed default realm to debug
- Changed default display limit for variants in column dialog to 15
- Removed apply filter button -> automatically load filter when changing
- Use user (id) as base name for storing files
- Rename ID -> Transcript
- Remove the "exon" part from the exon string on display
- Use createObjectURL to download files with JWT

### Removed
- Removed unauthenticated access in debug config

## 0.0.2

### Changed
- better error handling with detail from request body
- better detail modal like in GSVar
- renamed GSVar -> GSvar
- make table columns uniform
- don't use the VariantFilterRequest model in the controller anymore

## 0.0.1

### Added 
- add OpenAPI-based API
- serve vcf2gsvar and an_vep (megSAP) via API
- serve VcfFilterAnnotate via API
- add multi-stage UI (log in, select file, select filter)
- automatically annotate (using an_vep) and convert (using vcf2gsvar) VCF files
- add linting for frontend and API
- add tests for API




api/openapi_server/__main__.py
#!/usr/bin/env python3

import os

import connexion

from openapi_server import encoder
from flask import send_from_directory, g
from flask_cors import CORS


def main():
    app = connexion.App(__name__, specification_dir='./openapi/')
    app.app.json_encoder = encoder.JSONEncoder
    app.add_api('openapi.yaml', arguments={'title': 'ngs-bits'})
    app.app.config['UPLOAD_FOLDER'] = os.path.abspath(
        os.getenv('DATA', os.getcwd()))
    app.app.config['PRODUCTION'] = os.getenv('PRODUCTION', False)
    app.app.config['ALLOWED_EXTENSIONS'] = {'vcf', 'GSvar'}
    origins = os.getenv('CORS_ORIGINS', '').split(',')
    if not app.app.config['PRODUCTION']:
        origins.append('http://localhost:8080')
    # enable CORS for all
    CORS(app.app, origins=origins, supports_credentials=True)

    @app.app.teardown_appcontext
    def close_connection(exception):
        db = getattr(g, '_database', None)
        if db is not None:
            db.close()

    if not app.app.config['PRODUCTION']:
        @app.app.route('/')
        def serve_index():
            return send_from_directory(directory=os.path.join(os.getcwd(), 'dist'), filename='index.html')

        @app.app.route('/<path:path>')
        def serve_dist(path):
            return send_from_directory(directory=os.path.join(os.getcwd(), 'dist'), filename=path)

    app.run(port=os.getenv('PORT', 9000), debug=not app.app.config['PRODUCTION'], use_debugger=not app.app.config['PRODUCTION'],
            use_reloader=not app.app.config['PRODUCTION'], passthrough_errors=not app.app.config['PRODUCTION'])


if __name__ == '__main__':
    main()
